libraries/type-length-value/js/src/tlvState.ts

```diff
@@ -0,0 +1,132 @@
+import { TlvInvalidAccountSizeError, TlvInvalidDiscriminatorError } from './errors.js';
+
+export type TlvNumberSize = 1 | 2 | 4 | 8;
+
+export interface TlvDiscriminator {
+    bytes: Buffer;
+}
+
+export type TlvType = Buffer | Uint8Array | number | bigint | TlvDiscriminator;
+
+export class TlvState {
+    private readonly tlvData: Buffer;
+    private readonly typeSize: TlvNumberSize;
+    private readonly lengthSize: TlvNumberSize;
+
+    public constructor(buffer: Buffer, typeSize: TlvNumberSize = 2, lengthSize: TlvNumberSize = 2, offset: number = 0) {
+        this.tlvData = buffer.subarray(offset);
+        this.typeSize = typeSize;
+        this.lengthSize = lengthSize;
+    }
+
+    /**
+     * Get the raw tlv data
+     *
+     * @return the raw tlv data
+     */
+    public get data(): Buffer {
+        return this.tlvData;
+    }
+
+    private readTlvNumberSize<T>(size: TlvNumberSize, offset: number, constructor: (x: number | bigint) => T): T {
+        switch (size) {
+            case 1:
+                return constructor(this.tlvData.readUInt8(offset));
+            case 2:
+                return constructor(this.tlvData.readUInt16LE(offset));
+            case 4:
+                return constructor(this.tlvData.readUInt32LE(offset));
+            case 8:
+                return constructor(this.tlvData.readBigUInt64LE(offset));
+        }
+    }
+
+    private tlvDiscriminatorMatches(type: TlvType, offset: number): boolean {
+        switch (typeof type) {
+            case 'number':
+                return this.readTlvNumberSize(this.typeSize, offset, Number) === type;
+            case 'bigint':
+                return this.readTlvNumberSize(this.typeSize, offset, BigInt) === type;
+            case 'object':
+                if (type instanceof Buffer) {
+                    return this.tlvData.subarray(offset, offset + this.typeSize).equals(type);
+                }
+                if ('bytes' in type) {
+                    return this.tlvData.subarray(offset, offset + this.typeSize).equals(type.bytes);
+                }
+                throw new TlvInvalidDiscriminatorError();
+            default:
+                throw new TlvInvalidDiscriminatorError();
+        }
```

Comment:
Let's rename this error to be `TlvTypeNotFoundError`

https://github.com/solana-labs/solana-program-library/blob/c2c516f6584f2c685e8e69eed5b3adbbf7fe5121/libraries/type-length-value/src/state.rs#L60-L64